src/function/gds/output_writer.cpp

```diff
@@ -104,99 +107,139 @@ void PathsOutputWriter::write(processor::FactorizedTable& fTable, nodeID_t dstNo
-                writePath(curPath);
-                fTable.append(vectors);
-            }
+        if (getTop(curPath)->getIter() == 1) {
+            writePath(curPath);
+            fTable.append(vectors);
             backtracking = true;
         }
         if (backtracking) {
-            // This code checks if we should switch from backtracking to forward-tracking, i.e.,
-            // moving forward in the DFS logic to find paths. We switch from backtracking if:
-            // (i) the current top element in the stack has a nextPtr, i.e., the top node has
-            // more parent edges in the BFS graph AND:
-            // (ii.1) if this is the first element in the stack (curPath.size() == 1), i.e., we
-            // are enumerating the parents of the destination, then we should switch to
-            // forward-tracking if the next parent has visited the destination at a length
-            // that's greater than or equal to the lower bound of the recursive join. Otherwise,
-            // we'll enumerate paths that are smaller than the lower; OR
-            // (ii.2) if this is not the first element in the stack, i.e., then we should switch
-            // to forward tracking only if the next parent of the top node in the stack has the
-            // same iter value as the current parent. That's because the levels/iter need to
-            // decrease by 1 each time we add a new node in the stack.
-            if (top->getNextPtr() != nullptr &&
-                ((curPath.size() == 1 && top->getNextPtr()->getIter() >= info.lowerBound) ||
-                    top->getNextPtr()->getIter() == top->getIter())) {
-                curPath[curPath.size() - 1] = top->getNextPtr();
-                backtracking = false;
+            auto next = getTop(curPath)->getNextPtr();
+            while (true) {
+                if (!canNextReplacePathTop(next, curPath)) {
+                    curPath.pop_back();
+                    break;
+                }
+                // Further check next against path node mask (predicate).
+                if (!checkNewPathElement(curPath, next)) {
+                    next = next->getNextPtr();
+                    continue;
+                }
+                // Next is a valid path element. Push into stack and switch to forward track.
+                curPath[curPath.size() - 1] = next;
                 if (curPath.size() == 1) {
                     setLength(lengthVector.get(), curPath[0]->getIter());
                 }
-            } else {
-                curPath.pop_back();
+                backtracking = false;
+                break;
             }
         } else {
-            auto parent = bfsGraph.getInitialParentAndNextPtr(topNodeID);
-            while (parent->getIter() != top->getIter() - 1) {
+            auto top = getTop(curPath);
+            auto parent = bfsGraph.getInitialParentAndNextPtr(top->getNodeID());
+            while (true) {
```

Comment:
Isn't this while loop redundant too? This while (true) loop is inside the following broader code structure:
```
while (!curPath.empty()) {
   ...
   if (backtracking) {
       ...
   } else {
      .... // your while (true) loop is here
   }
}
```
So if you don't add the while (true) loop in that else branch, if you remove it, the above while loop will still loop. So let's remove it to simplify the code.

Reply:
I don't think this is the case. This for loop is trying to get the parent of parent. If we remove the while loop directly, we will always access the parent of top. So it's not as simple as removing a while loop. Also the previous version is doing something very similar

```
while (!curPath.empty) {
      if (backtracking) {
      } else {
                auto parent = bfsGraph.getParentListHead(topNodeID);
                while (parent->getIter() != top->getIter() - 1) {
                    parent = parent->getNextPtr();
                }
      }
}
```

So I would not change.